internal/services/scheduler/v1/scheduler.go

```diff
@@ -699,18 +833,428 @@ func (s *Scheduler) notifyAfterConcurrency(ctx context.Context, tenantId string,
+			Str("batch_key", key.BatchKey).
+			Int("original_group_size", len(group)).
+			Int("deduped_group_size", len(dedupedGroup)).
+			Msg("prepared batch dispatcher group")
+
+		group = dedupedGroup
+
+		if len(group) == 0 {
+			continue
+		}
+
+		meta := group[0].Batch
+		if meta == nil {
+			meta = &repov1.BatchAssignmentMetadata{}
+		}
+
+		batchID := strings.TrimSpace(meta.BatchID)
+		if batchID == "" {
+			batchID = uuid.NewString()
+		}
+
+		shouldRelease := meta.MaxRuns > 0 && strings.TrimSpace(key.BatchKey) != "" && batchID != ""
+		releaseOnError := func() {
+			if !shouldRelease {
+				return
+			}
+
+			if err := s.repov1.Tasks().DeleteTaskBatchRun(ctx, tenantId, batchID); err != nil {
+				s.l.Error().
+					Err(err).
+					Str("tenant_id", tenantId).
+					Str("batch_id", batchID).
+					Msg("failed to release batch reservation after error")
+			}
+		}
+
+		dispatcherID, ok := workerIdToDispatcherId[key.WorkerID]
+		if !ok {
+			s.internalRetry(ctx, tenantId, group...)
+			releaseOnError()
+			result = multierror.Append(result, fmt.Errorf("could not assign batch to worker %s: dispatcher id missing", key.WorkerID))
+			continue
+		}
+
+		triggeredAt := meta.TriggeredAt
+		if triggeredAt.IsZero() {
+			triggeredAt = time.Now().UTC()
+		}
+
+		batchSize := len(group)
+		configuredSize := int(meta.ConfiguredBatchSize)
+		if configuredSize <= 0 {
+			configuredSize = batchSize
+		}
+
+		flushReason := describeBatchFlushReason(meta.Reason, configuredSize, time.Duration(meta.ConfiguredFlushIntervalMs)*time.Millisecond)
+
+		assignmentsPayload := make([]repov1.TaskBatchAssignment, 0, len(group))
+		taskIds := make([]int64, 0, len(group))
+
+		for idx, item := range group {
+			if item == nil || item.QueueItem == nil || !item.QueueItem.TaskInsertedAt.Valid {
+				continue
+			}
+
+			assignmentsPayload = append(assignmentsPayload, repov1.TaskBatchAssignment{
+				TaskID:         item.QueueItem.TaskID,
+				TaskInsertedAt: item.QueueItem.TaskInsertedAt.Time,
+				BatchIndex:     idx,
+			})
+
+			taskIds = append(taskIds, item.QueueItem.TaskID)
+		}
+
+		if len(assignmentsPayload) == 0 {
+			releaseOnError()
+			continue
+		}
+		if err := s.repov1.Tasks().UpdateTaskBatchMetadata(ctx, tenantId, batchID, key.WorkerID, key.BatchKey, batchSize, assignmentsPayload); err != nil {
+			s.internalRetry(ctx, tenantId, group...)
+			releaseOnError()
+			result = multierror.Append(result, fmt.Errorf("could not persist batch metadata: %w", err))
+			continue
+		}
+
+		startPayload := tasktypes.StartBatchTaskPayload{
+			TenantId:      tenantId,
+			WorkerId:      key.WorkerID,
+			ActionId:      key.ActionID,
+			BatchId:       batchID,
+			ExpectedSize:  batchSize,
+			BatchKey:      key.BatchKey,
+			TriggerReason: flushReason,
+			TriggerTime:   triggeredAt,
+		}
+
+		if meta.MaxRuns > 0 {
+			maxRuns := int(meta.MaxRuns)
+			startPayload.MaxRuns = &maxRuns
+		}
+
+		startMsg, err := tasktypes.StartBatchMessage(tenantId, startPayload)
```

Comment:
What happens if this message is received by the dispatcher after it gets one of the individual tasks? Ordering is not guaranteed on the dispatcher side because of a QOS which is greater than 1, so if the worker expects this message to be received strictly before all of the batch's task assignments, this might not work.